POST: storing and using httr2 cache tokens on CI (#1173)
ropensci/roweb3: content/blog/2025-10-30-httr2-cache/index.md

---

> title: A Primer on Domain Verification

Author: lol

This issue popped up as we were working on a refactoring of the `meetupr` package, which interfaces with the Meetup API.
We were switching from the older `httr` package to `httr2`, which has a more modern and flexible approach to OAuth.
We also knew that we wanted to support CI/CD for users of the package, as R-Ladies uses it to store an archive of Meetup groups and events.

This post walks through a general pattern for solving this problem, using examples from the recent new development on the `meetupr` package (which wraps the Meetup API) to illustrate the concepts.

## Understanding httr2's OAuth Flow

httr2's `req_oauth_auth_code()` handles OAuth2 authentication for local development:

1. Opens the user's browser for authentication  
2. Exchanges the authorization code for tokens (access + refresh)  
3. Caches the tokens to disk for future use  
4. Automatically refreshes expired tokens using the refresh token  

The cached token file contains everything needed: the access token, refresh token, and expiration time. 

> This means httr2 handles token refresh transparently - no additional code required.

Author: I was trying to elaborate on it, but agree its repetitive.

## The CI Problem

In CI environments, this interactive flow fails because:

1. **No browser**: CI environments can't open browsers for user authentication  
2. **Fresh environments**: Each CI run starts with no cached credentials  

### What About httr2's Non-Interactive OAuth Methods?

httr2 provides several non-interactive authentication methods that work well for CI:

- `req_oauth_client_credentials()`: For service-to-service authentication  
- `req_oauth_bearer_jwt()`: For JWT-based authentication  
- `req_oauth_device()`: For device flow authentication  

These are excellent options when available! 
However, they have limitations:

- **Not universally supported**: Many APIs don't support device flow or service accounts  
- **Account restrictions**: Some providers (like Meetup) restrict certain OAuth methods to paid accounts  
- **Different permission models**: Service accounts often have different access levels than user accounts  

In the case of `meetupr`, we discovered that:

- JWT authentication requires a Pro Meetup account  
- OAuth with localhost redirect requires a Pro Meetup account  
- Device flow is not supported by the Meetup API  

> - Only Pro accounts can create new OAuth applications at all (previously made applications are still valid, though) 

Author: I did use it originally, and thought maybe it was too unfamiliar!

```

**CI configuration:**
Set three secrets in your CI provider:
- `API_TOKEN`: The base64-encoded token content
- `API_TOKEN_FILENAME`: The base64-encoded filename
- `HTTR2_OAUTH_CACHE=.` as environment variable

**In CI, your code works normally:**
```r
library(yourpackage)
result <- your_api_function()  # Just works!
```

## Why This Approach Works

1. **httr2 does the heavy lifting**: Token refresh, expiration handling, and caching are all built-in  
2. **Refresh tokens enable long-term access**: Even though access tokens expire (often hourly), refresh tokens typically last weeks or months  
3. **No duplicate logic**: You're not reimplementing OAuth for CI  
4. **Same code everywhere**: No conditional logic for local vs CI environments  
5. **Secure by default**: CI secrets are encrypted at rest; base64 encoding is sufficient for transfer  

## Conclusion

Managing OAuth in CI doesn't require complex custom logic. 
When httr2's purpose-built non-interactive methods aren't available due to API provider limitations, transferring cached tokens provides a robust alternative.

By understanding how httr2 caches tokens you can create a solution that:

> - Minimizes package code  

Author: other solutions would need much more code, but that's beside the point here.

---
title: "Cache Me If You Can – httr2 OAuth Token Management for CI/CD"
author:
- Athanasia Mo Mowinckel

> - Maëlle Salmon

Reviewer: I wonder whether I should be an editor instead, given what my input was.

Author: I leave that to you, but I think of you as co-auther if that counts for something.

editor:
  -
date: '2026-03-30'
slug: httr2-cache
description: Learn how to manage httr2 OAuth tokens in CI/CD, where interactive authentication is not possible.
output: hugodown::md_document
tags:
  - oAuth
  - tech notes
crossposts:
- name: R-Ladies blog blog
  url: https://rladies.org/blog/2025/httr2-cache/
params:
  doi: ""
---

> We ran into this problem while refactoring the [meetupr](https://github.com/rladies/meetupr) package, which wraps the Meetup API and is maintained by the R-Ladies organization.

Reviewer: make the problem explicit even if that's the title of the post?

> The package was switching from the superseded httr to httr2, which opens a browser for user authentication and makes it clear what permissions you're granting.

Reviewer: didn't httr do that?

Author: yes it did. Just saved it differently. altered the language now :)

> The package was switching from the superseded httr to httr2, which opens a browser for user authentication and makes it clear what permissions you're granting.

Reviewer: ```suggestion
The package was switching from the superseded [httr](https://httr.r-lib.org/) to [httr2](https://httr2.r-lib.org/), which opens a browser for user authentication and makes it clear what permissions you're granting.
```

R-Ladies uses meetupr in CI to archive Meetup groups and events, so we needed authentication that works without a human sitting at a keyboard.
httr2 handles interactive OAuth well, but CI environments don't have browsers.
This post walks through how we solved it — and the general pattern is useful for any R package wrapping an OAuth API.

## Understanding httr2's OAuth flow

> httr2's `req_oauth_auth_code()` handles [OAuth2](https://blog.r-hub.io/2021/01/25/oauth-2.0/) authentication for local development:

Reviewer: I obviously like that R-hub post but maybe link to https://httr2.r-lib.org/articles/oauth.html instead/as well?

1. Opens the user's browser for authentication    
2. Exchanges the authorization code for tokens (access + refresh)    
3. Encrypts the token and caches it to disk (by default under the httr2 cache directory)    
4. Automatically refreshes expired tokens using the refresh token    

That last point matters.
Access tokens are short-lived — often an hour — but refresh tokens can last weeks or months.

> As long as you have a valid refresh token, you can keep getting fresh access tokens without the browser dance.

Reviewer: I will forever by mystified by this. Why is it safer

Author: I know. I get very confused by this....

## The CI problem

CI environments break this flow in two ways:

1. No browser available for the initial authentication    
2. Fresh environments every run — no cached credentials    

### What about httr2's non-interactive methods?

httr2 provides [several non-interactive authentication methods](https://httr2.r-lib.org/reference/req_oauth_auth_code.html#see-also) — device flow, service accounts, JWT.
These are the right first choice when your API supports them.

But not every API does.
In the case of Meetup, we discovered that:

> - JWT authentication requires a Pro Meetup account    

Reviewer: I feel like adding some sort of money emoji. 💸 💰 🤑

> - Creating new OAuth applications requires a Pro account (existing apps are [grandfathered](https://en.wikipedia.org/wiki/Grandfather_clause))    

Reviewer: ```suggestion
- Creating new OAuth applications also requires a Meetup Pro account (existing apps are [grandfathered](https://en.wikipedia.org/wiki/Grandfather_clause))    
```

- Device flow is not supported at all    

> So we needed two paths: JWT for Pro account holders, and something else for everyone with a grandfathered OAuth app.

Reviewer: Add the caveat that it means this thing won't work for other users?

## The solution: encrypted token files

The approach we landed on has three tiers, tried in order:

1. _JWT_ — for Pro accounts that support it    
2. _Encrypted token file_ — for regular OAuth apps in CI    
3. _Interactive OAuth_ — the standard browser flow for local development    

The request authentication function tries each method in sequence:

```r
req_auth <- function(req, client_name, cache = TRUE, ...) {
  # Custom function that checks what authentication is available
  auth <- meetupr_auth_status(silent = TRUE)

  if (auth$jwt$available) {
    # JWT bearer token (Pro accounts)

>     return(req |> httr2::req_oauth_bearer_jwt(

Reviewer: could the pipeline be assigned to a variable then return would return this variable? I find this a bit hard to read. 💅

      client = meetupr_client(...),
      claim = httr2::jwt_claim(...)
    ))
  }

  # Try encrypted token file
  token <- tryCatch(
    meetupr_encrypt_load(client_name = client_name),
    error = function(e) NULL
  )

  if (!is.null(token)) {
    return(httr2::req_auth_bearer_token(req, token))
  }

  # Fall back to interactive browser flow
  req |> httr2::req_oauth_auth_code(
    client = meetupr_client(...),
    auth_url = meetupr_api_urls()$auth,
    cache_disk = cache
  )
}
```

JWT is straightforward — if you have the credentials, httr2 handles the rest.

> The interesting part is the encrypted token path.

Reviewer: ```suggestion
The interesting part is the encrypted token path.
The one we wanted to use in our CI workflows!
```

### How it works

The idea is simple: authenticate locally once, then encrypt and commit the token so CI can use it.
Only the decryption password goes into CI secrets — the encrypted file itself lives in your repository.

We use the [cyphr](https://cran.r-project.org/package=cyphr) and [sodium](https://cran.r-project.org/package=sodium) packages for encryption.

> The token object is saved as an RDS file, then encrypted with a randomly generated password using NaCl symmetric encryption.

Reviewer: link?

mention security, e.g. how difficult would it be to crack it?

Here's the setup function, simplified for clarity:

```r
meetupr_encrypt_setup <- function(path = ".meetupr.rds", password = NULL) {
  token <- httr2::oauth_token_cached(
    client = meetupr_client(),
    flow = httr2::oauth_flow_auth_code,
    flow_params = meetupr_oauth_flow_params()
  )

  if (is.null(password)) {
    password <- sodium::bin2hex(sodium::random(32))
  }

  key <- cyphr::key_sodium(sodium::hex2bin(password))
  tmp <- tempfile(fileext = ".rds")
  saveRDS(token, tmp)
  cyphr::encrypt_file(tmp, key = key, dest = path)

  cli::cli_bullets(c(
    "v" = "Encrypted token saved to {.path {path}}",
    "i" = "Commit this file to your repository",
    "!" = "Add encryption password to CI secrets"
  ))

  invisible(password)
}
```

The user runs `meetupr_auth()` to authenticate in the browser, then `meetupr_encrypt_setup()` to create the encrypted file.

> They commit that file and add the password as a CI secret.

Reviewer: ```suggestion
They commit and push that file. 
Then they add the password for deencryption as a CI secret.
```

Link to GitHub docs for CI secrets?

1. Decrypts the token file using the password from the CI secret    
2. Checks if the access token has expired    
3. If expired, refreshes it using the refresh token and saves the rotated encrypted file back    

```r
meetupr_encrypt_load <- function(path, password) {
  key <- cyphr::key_sodium(sodium::hex2bin(password))
  tmp <- tempfile(fileext = ".rds")
  cyphr::decrypt_file(path, key = key, dest = tmp)
  token <- readRDS(tmp)

  token_age <- Sys.time() - as.POSIXct(
    token$expires_at, origin = "1970-01-01"
  )
  if (token_age < 0) return(token)

  new_token <- refresh_oauth_token(token)
  saveRDS(new_token, tmp)
  cyphr::encrypt_file(tmp, key = key, dest = path)
  new_token
}
```

The rotation step is what makes this approach durable.
Instead of the encrypted file going stale after the access token expires, it stays current.
The caveat being, that every token refresh needs to be committed and pushed to the repo by the CI.

### What we tried first

Our first approach — suggested by [Noam Ross](https://www.noamross.net/) on the rOpenSci Slack — was to base64-encode the raw httr2 cache file and store it as a CI environment variable.

> This worked, but had drawbacks: environment variables have size limits on some CI providers, you had to store both the token content _and_ httr2's hash-based filename as separate secrets, and there was no natural way to write back a rotated token.

Reviewer: Isn't it weird that there's no natural way to do that?

Author: maybe? I kind of felt it was meaningful that that specific thing is hard to do?

      - name: Install dependencies
        run: |
          install.packages('pak')
          pak::pak('rladies/meetupr')
        shell: Rscript {0}

      - name: Load encrypted token and refresh
        run: |
          meetupr::meetupr_encrypt_load()
        shell: Rscript {0}

      - name: Run your script
        run: Rscript your_script.R

      - name: Commit and push updated token
        run: |
          git config user.name "github-actions[bot]"
          git config user.email "github-actions[bot]@users.noreply.github.com"
          git add .meetupr.rds
          git commit -m "Rotate meetupr token [skip ci]" || echo "No changes to commit"
          git push
```

For Pro accounts using JWT, `use_gha_jwt_token()` generates a simpler workflow that injects JWT credentials as environment variables.
You can see an example of this in the [R-Ladies meetup archive repository](https://github.com/rladies/meetup_archive/actions/runs/23193611609/workflow.)

## Complete user workflow

From a meetupr user's perspective:

_One-time local setup:_

```r
library(meetupr)

meetupr_auth()

password <- meetupr_encrypt_setup()
# → Encrypted token saved to .meetupr.rds

> # → Commit this file to your repository

Reviewer: random thought: could it do the committing like usethis' functions do, after asking?

Author: probably, but I kind of didnt want to build that in to allow users more flexibility in what they do.

# → Add encryption password to CI secrets
```

_Repository setup:_

1. Commit `.meetupr.rds` to your repository    
2. Add the encryption password as a CI secret named `ENCRYPT_PWD`    

> 3. Run `meetupr::use_gha_encrypted_token()` to create the workflow file    

Reviewer: ```suggestion
3. Run `meetupr::use_gha_encrypted_token()` to create the workflow file, commit it to your repository
```

result <- get_events("R-Ladies-Global")
```

The authentication happens transparently — `meetupr_encrypt_load()` is called internally when the package builds its API request.

## The general pattern

The specifics here are meetupr's, but the pattern generalizes to any R package wrapping an OAuth API that needs CI support:

1. Authenticate interactively once and capture the httr2 token object    
2. Encrypt it with cyphr/sodium and commit the encrypted file    
3. In CI, decrypt, check expiration, refresh if needed, re-encrypt    
4. Store only the decryption password as a CI secret    

The encrypted file in the repository acts as a renewable credential.
Each CI run that refreshes the token writes back the updated version, so the refresh token stays current.
If the refresh token does eventually expire — say the user revokes access or the API provider has a short refresh window — the user re-runs the setup locally.

## Security considerations

The encrypted token file is safe to commit because it uses NaCl symmetric encryption via sodium.
Without the password, the file is opaque.

The password itself lives only in CI secrets, which are encrypted at rest and not exposed in logs.

> The token file contains OAuth tokens — not user credentials — so the blast radius of a compromised token is limited to the API permissions granted to that OAuth app.

Reviewer: ```suggestion
The token file contains OAuth tokens — not user credentials — so the [blast radius](https://en.wikipedia.org/wiki/Blast_radius#Cloud_Computing) of a compromised token is limited to the API permissions granted to that OAuth app.
```

> 

Reviewer: Also mention you could make the GH repo private for an added layer?

> For sensitive environments, the weekly rotation workflow keeps the token fresh and gives you an audit trail of when tokens were last refreshed.

Reviewer: I'm not sure I follow how that helps?

Author: ```suggestion
```

I dont either really...

output: hugodown::md_document
tags:
  - oAuth
  - tech notes
crossposts:
- name: R-Ladies blog post
  url: https://rladies.org/blog/2026/httr2-cache/
params:
  doi: ""
---

How do you authenticate with an API from CI/CD (continuous integration/continuous deployment) when the authentication flow requires a human to click "Allow" in a browser?

We ran into this problem while refactoring the [meetupr](https://github.com/rladies/meetupr) package, which wraps the Meetup API and is maintained by the R-Ladies organization.
The package was switching from the superseded [httr](https://httr.r-lib.org/) to [httr2](https://httr2.r-lib.org/).
Both packages handle OAuth by directing the user to a browser where they grant permissions — httr2 just does it with a more modern and flexible approach.

R-Ladies uses meetupr in CI to archive Meetup groups and events, so we needed authentication that works without a human sitting at a keyboard.
httr2 handles interactive OAuth well, but CI environments don't have browsers.
This post walks through how we solved it — and the general pattern is useful for any R package wrapping an OAuth API.

## Understanding httr2's OAuth flow

> httr2's `req_oauth_auth_code()` handles [OAuth2](https://blog.r-hub.io/2021/01/25/oauth-2.0/) authentication for local development:

Author: ```suggestion
```

1. Opens the user's browser for authentication
2. Exchanges the authorization code for tokens (access + refresh)
3. Encrypts the token and caches it to disk (under `tools::R_user_dir("httr2", "cache")` by default)
4. Automatically refreshes expired tokens using the refresh token

Access tokens are short-lived — often an hour — but refresh tokens can last weeks or months.

> As long as you have a valid refresh token, you can keep getting fresh access tokens without the browser dance.

Author: ```suggestion
As long as you have a valid refresh token, you can keep getting fresh access tokens without the browser dance.
Even more information on oAuth 2.0 in [this R-hub post](https://blog.r-hub.io/2021/01/25/oauth-2.0/)
```

## The CI problem

CI environments break this flow in two ways:

1. No browser available for the initial authentication
2. Fresh environments every run — no cached credentials

### What about httr2's non-interactive methods?

httr2 provides [several non-interactive authentication methods](https://httr2.r-lib.org/reference/req_oauth_auth_code.html#see-also) — device flow, service accounts, JSON Web Tokens (JWT).
These are the right first choice when your API supports them.

But not every API does.
In the case of Meetup, we discovered that:

- JWT authentication requires a Pro Meetup account  
- Creating new OAuth applications also requires a Meetup Pro account (existing apps are [grandfathered](https://en.wikipedia.org/wiki/Grandfather_clause))  
- Device flow is not supported at all  

> So we needed two paths: JWT for Pro account holders, and something else for everyone with a grandfathered OAuth app.

Author: ```suggestion
So we needed two paths: JWT for Pro account holders, and something else for everyone with a grandfathered OAuth app, and for those without even that!
```

Author: ```suggestion
So we needed two paths: JWT for Pro account holders, and something else for everyone with a grandfathered OAuth app.
```

## The solution: encrypted token files

The approach we landed on has three tiers, tried in order:

1. _JWT_ — for Pro accounts that support it
2. _Encrypted token file_ — for regular OAuth apps in CI
3. _Interactive OAuth_ — the standard browser flow for local development

![Flowchart showing the authentication decision tree. Starting from an API request, the package first checks if JWT credentials are available. If yes, it uses JWT bearer token auth. If not, it checks for an encrypted token file. If found, it decrypts and uses that token. If neither is available, it falls back to the interactive browser OAuth flow.](auth-decision-flow.png "Authentication decision flow — the package tries each method in order and uses the first one available.")

The request authentication function tries each method in sequence:

```r
req_auth <- function(req, client_name, cache = TRUE, ...) {
  # Custom function that checks what authentication is available
  auth <- meetupr_auth_status(silent = TRUE)

  if (auth$jwt$available) {
    # JWT bearer token (Pro accounts)
    return(req |> httr2::req_oauth_bearer_jwt(
      client = meetupr_client(...),
      claim = httr2::jwt_claim(...)

>     ))

Author: ```suggestion
  req <- req |> 
  httr2::req_oauth_bearer_jwt(
        client = meetupr_client(...),
        claim = httr2::jwt_claim(...)
      ))
    return(req)
```

---

> title: "Cache Me If You Can – httr2 OAuth Token Management for CI/CD"

Reviewer 2: Great title!

author:
- Athanasia Mo Mowinckel

> - Maëlle Salmon

Reviewer 2: ```suggestion
- Maëlle Salmon
editor: Steffi LaZerte
```

> date: '2026-03-30'

Reviewer 2: Remember to adjust the date here and in the folder name

slug: httr2-cache
description: Learn how to manage httr2 OAuth tokens in CI/CD, where interactive authentication is not possible.
output: hugodown::md_document
tags:
  - oAuth

>   - tech notes

Reviewer 2: ```suggestion
  - tech notes
  - CI
  - API
```

crossposts:

> - name: R-Ladies blog post

Reviewer 2: The text becomes "This is cross-posted on xxx" so leaving out 'post' sounds a bit better
```suggestion
- name: R-Ladies blog
```

  url: https://rladies.org/blog/2026/httr2-cache/
params:

>   doi: ""

Reviewer 2: ```suggestion
  doi: "10.59350/c18c1-fv109"
```

1. No browser available for the initial authentication
2. Fresh environments every run — no cached credentials

### What about httr2's non-interactive methods?

httr2 provides [several non-interactive authentication methods](https://httr2.r-lib.org/reference/req_oauth_auth_code.html#see-also) — device flow, service accounts, JSON Web Tokens (JWT).
These are the right first choice when your API supports them.

But not every API does.
In the case of Meetup, we discovered that:

- JWT authentication requires a Pro Meetup account :money_mouth_face:
- Creating new OAuth applications also requires a Meetup Pro account (existing apps are [grandfathered](https://en.wikipedia.org/wiki/Grandfather_clause)) :moneybag:  
- Device flow is not supported at all  

So we needed two paths: JWT for Pro account holders, and something else for everyone with a grandfathered OAuth app, and for those without even that!

## The solution: encrypted token files

The approach we landed on has three tiers, tried in order:

1. _JWT_ — for Pro accounts that support it
2. _Encrypted token file_ — for regular OAuth apps in CI
3. _Interactive OAuth_ — the standard browser flow for local development

> ![Flowchart showing the authentication decision tree. Starting from an API request, the package first checks if JWT credentials are available. If yes, it uses JWT bearer token auth. If not, it checks for an encrypted token file. If found, it decrypts and uses that token. If neither is available, it falls back to the interactive browser OAuth flow.](auth-decision-flow.png "Authentication decision flow — the package tries each method in order and uses the first one available.")

Reviewer 2: Should the "\n" be New Lines? Looks like that didn't work in the figure...

The request authentication function tries each method in sequence:

```r
req_auth <- function(req, client_name, cache = TRUE, ...) {
  # Custom function that checks what authentication is available
  auth <- meetupr_auth_status(silent = TRUE)

  if (auth$jwt$available) {
    # JWT bearer token (Pro accounts)

>   req <- req |> 
>   httr2::req_oauth_bearer_jwt(
>         client = meetupr_client(...),
>         claim = httr2::jwt_claim(...)
>       ))
>     return(req)

Reviewer 2: Just to match the rest
```suggestion
  req <- req |> 
    httr2::req_oauth_bearer_jwt(
      client = meetupr_client(...),
      claim = httr2::jwt_claim(...)
    ))
  return(req)
```

  }

  # Try encrypted token file
  token <- tryCatch(
    meetupr_encrypt_load(client_name = client_name),
    error = function(e) NULL
  )

  if (!is.null(token)) {
    return(httr2::req_auth_bearer_token(req, token))
  }

  # Fall back to interactive browser flow
  req |> httr2::req_oauth_auth_code(
    client = meetupr_client(...),
    auth_url = meetupr_api_urls()$auth,
    cache_disk = cache
  )
}
```

JWT is straightforward — if you have the credentials, httr2 handles the rest.
The interesting part is the encrypted token path.
The one we wanted to use in our CI workflows!

> 

Reviewer 2: I suggest moving the "What we tried first" section here, and renaming it "Why encrypted tokens?" (see my other comments below)

### How it works

> The idea is simple: authenticate locally once, then encrypt and commit the token so CI can use it.
> Only the decryption password goes into CI secrets — the encrypted file itself lives in your repository.

Reviewer 2: If you chose to move the "What we tried first" section to just above, I also suggest adding this first paragraph to the end of that section (see my suggestion, below), and then you then don't need the paragraph here. 

```suggestion
```

On the CI side, `meetupr_encrypt_load()` does three things:

1. Decrypts the token file using the password from the CI secret
2. Checks if the access token has expired
3. If expired, refreshes it using the refresh token and saves the rotated encrypted file back

```r
meetupr_encrypt_load <- function(path, password) {
  key <- cyphr::key_sodium(sodium::hex2bin(password))
  tmp <- tempfile(fileext = ".rds")
  cyphr::decrypt_file(path, key = key, dest = tmp)
  token <- readRDS(tmp)

  token_age <- Sys.time() - as.POSIXct(
    token$expires_at, origin = "1970-01-01"
  )
  if (token_age < 0) return(token)

  new_token <- refresh_oauth_token(token)
  saveRDS(new_token, tmp)
  cyphr::encrypt_file(tmp, key = key, dest = path)
  new_token
}
```

The rotation step is what makes this approach durable.
Instead of the encrypted file going stale after the access token expires, it stays current.
The caveat being, that every token refresh needs to be committed and pushed to the repo by the CI.

> ### What we tried first

Reviewer 2: I wonder if this section should go right before you describe the encryption part (i.e. before the 'How it works' part). It seems to break the flow a bit here and seems out of order to talk about what you did first after you talk about what you did second 😁

If you do that, consider remaming this section to "Why encrypted tokens?"

> Our first approach — suggested by [Noam Ross](https://www.noamross.net/) on the rOpenSci Slack — was to base64-encode the raw httr2 cache file and store it as a CI environment variable.

Reviewer 2: ```suggestion
Our first approach — suggested by [Noam Ross](/author/noam-ross/) on the rOpenSci Slack — was to base64-encode the raw httr2 cache file and store it as a CI environment variable.
```

This worked, but had drawbacks: environment variables have size limits on some CI providers, you had to store both the token content _and_ httr2's hash-based filename as separate secrets, and there was no natural way to write back a rotated token.

> It is a bit surprising! 

Reviewer 2: Was it really surprising? The next sentence seems to argue that this is expected? Perhaps either remove the "It is a bit surprising!" Statement, or tweak it to something like "It was a bit surprising, but made sense once we thought about it a bit more." (or similar).

Many CI/CD systems are designed for stateless, ephemeral runs, so persisting and updating secrets (like OAuth tokens) isn't a built-in feature. 
Most CI providers treat secrets as write-only environment variables, not as files that can be updated and committed back. 
This makes token rotation awkward: you can't just "update the secret" from within the workflow.

> The encrypted file approach is simpler for users — commit one file, set one secret — and gives CI a place to write back the refreshed token.
> By using an encrypted file in the repository, you sidestep this limitation. 
> The file acts as a renewable credential that CI can update and push, while the decryption password remains protected as a CI secret.
> This pattern is not common in most documentation, but it works well for R packages and similar use cases.

Reviewer 2: If moving this section to earlier, consider adding a bit of the "How it works" introduction here so it flows directly into that section.

```suggestion
So we thought of another approach, using encrypted token files. The idea is simple: authenticate locally once, then encrypt and commit the token so CI can use it. Only the decryption password goes into CI secrets — the encrypted file itself lives in your repository. 
This is simpler for users — commit one file, set one secret — and gives CI a place to write back the refreshed token.
By using an encrypted file in the repository, you sidestep the limitations of CI environmental variables. 
The file acts as a renewable credential that CI can update and push, while the decryption password remains protected as a CI secret.
This pattern is not common in most documentation, but it works well for R packages and similar use cases.
```

> ## CI

Reviewer 2: We've been talking about CI a lot, so I wasn't sure what to expect with the heading "CI". Consider "CI maintenance", or perhaps "CI token maintenance" Or "Token maintenance on CI"
```suggestion
## Token maintenance on CI
```

> meetupr provides helper functions that generate GitHub Actions workflow files.

Reviewer 2: Remind us :)
```suggestion
The meetupr package provides helper functions that generate GitHub Actions workflow files.
```

> For encrypted token rotation:

Reviewer 2: ```suggestion
For encrypted token rotation we added a new function:
```

```yaml
name: Meetupr Token Rotation

on:
  push:
    branches: [main]
  workflow_dispatch:
  schedule:
    - cron: "0 0 * * 1"

jobs:
  rotate-token:
    runs-on: ubuntu-latest
    env:
      meetupr_encrypt_pwd: ${{ secrets.ENCRYPT_PWD }}
      meetupr_encrypt_path: .meetupr.rds

    steps:
      - uses: actions/checkout@v4

      - name: Set up R
        uses: r-lib/actions/setup-r@v2

      - name: Install dependencies
        run: |
          install.packages('pak')
          pak::pak('rladies/meetupr')
        shell: Rscript {0}

      - name: Load encrypted token and refresh
        run: |
          meetupr::meetupr_encrypt_load()
        shell: Rscript {0}

      - name: Run your script
        run: Rscript your_script.R

      - name: Commit and push updated token
        run: |
          git config user.name "github-actions[bot]"
          git config user.email "github-actions[bot]@users.noreply.github.com"
          git add .meetupr.rds
          git commit -m "Rotate meetupr token [skip ci]" || echo "No changes to commit"
          git push
```

> For Pro accounts using JWT, `use_gha_jwt_token()` generates a simpler workflow that injects JWT credentials as environment variables.
> You can see an example of this in the [R-Ladies meetup archive repository](https://github.com/rladies/meetup_archive/actions/runs/23193611609/workflow.)

Reviewer 2: ```suggestion
For Pro accounts using JWT, `meetupr::use_gha_jwt_token()` generates a simpler workflow that injects JWT credentials as environment variables.
You can see an example of this in the [R-Ladies meetup archive repository](https://github.com/rladies/meetup_archive/actions/runs/23193611609/workflow).
```

## Complete user workflow

> From a meetupr user's perspective:

Reviewer 2: It's probably just stylistic preferences, but I found the nested labels (i.e. the two statements ending with ':' right at the start), a bit confusing, so I suggest rewording, so it just one series of steps. 
```suggestion
If you are a meetupr user, you would use the following workflow.
```

> _One-time local setup:_

Reviewer 2: ```suggestion
One-time local setup:
```

```r
library(meetupr)

meetupr_auth()

password <- meetupr_encrypt_setup()
# → Encrypted token saved to .meetupr.rds
# → Commit this file to your repository
# → Add encryption password to CI secrets
```

> _Repository setup:_

Reviewer 2: ```suggestion
Repository setup:
```

1. Commit `.meetupr.rds` to your repository  
2. Add the encryption password as a CI secret named `ENCRYPT_PWD`  
3. Run `meetupr::use_gha_encrypted_token()` to create the workflow file, commit it to your repository  

> _In CI, your code works normally:_

Reviewer 2: ```suggestion
In CI, your code works normally:
```

```r
library(meetupr)
result <- get_events("R-Ladies-Global")
```

The authentication happens transparently — `meetupr_encrypt_load()` is called internally when the package builds its API request.

> ## The general pattern

Reviewer 2: Maybe? What do you think?
```suggestion
## General applicability
```

> The specifics here are meetupr's, but the pattern generalizes to any R package wrapping an OAuth API that needs CI support:

Reviewer 2: ```suggestion
The specifics we demonstrate in this post reflect meetupr's setup, but the pattern generalizes to any R package wrapping an OAuth API that needs CI support:
```

1. Authenticate interactively once and capture the httr2 token object
2. Encrypt it with cyphr/sodium and commit the encrypted file
3. In CI, decrypt, check expiration, refresh if needed, re-encrypt
4. Store only the decryption password as a CI secret

The encrypted file in the repository acts as a renewable credential.
Each CI run that refreshes the token writes back the updated version, so the refresh token stays current.
If the refresh token does eventually expire — say the user revokes access or the API provider has a short refresh window — the user re-runs the setup locally.

> ## Security considerations
> 

Reviewer 2: I only suggest removing this heading because I think it'll flow better if you add a concluding sentence or two at the end.
```suggestion

```

The encrypted token file is safe to commit because it uses NaCl symmetric encryption via sodium.
The password itself lives only in CI secrets, which are encrypted at rest and not exposed in logs.
The token file contains OAuth tokens — not user credentials — so the [blast radius](https://en.wikipedia.org/wiki/Blast_radius#Cloud_Computing) of a compromised token is limited to the API permissions granted to that OAuth app.

For an added layer of protection, you can also make your GitHub repository private. This ensures that even the encrypted token file is not publicly accessible, further reducing the risk of exposure. Only collaborators with access to the private repository would be able to obtain the encrypted file, while the decryption password remains secured in CI secrets.

> 

Reviewer 2: This paragraph definitely feels like a conclusion, but I think it could still use a sign off sentence or two.

Perhaps something simple such as "We hope this workflow helps others using OAuth APIs with CI support", and/or adding "Let us know how it works for you!" if you want to invite feedback.
